primitives/src/witness.rs

```diff
@@ -345,88 +356,160 @@ impl Decoder for WitnessDecoder {
+            // Short circuit for zero witness elements.
+            if witness_elements == 0 {
                 return Ok(false);
             }
 
-            // `cast_to_usize_if_valid` asserts length < 4,000,000, so no DoS vector here.
-            self.buffer = Vec::with_capacity(self.witness_elements);
+            // Allocate space for the index and buffer. The buffer
+            // is initialized to 128 bytes which should be large enough
+            // to cover most witnesses, the typical pubkey + signature
+            // and some overhead (e.g. P2WPKH witness is ~100 bytes),
+            // without reallocating.
+            let witness_index_space = witness_elements * 4;
+            self.cursor = witness_index_space;
+            self.content = alloc::vec![0u8; self.cursor + 128];
         }
 
+        let Some(witness_elements) = self.witness_elements else {
+            unreachable!("witness_elements must be Some after initial read")
+        };
+        let witness_index_space = witness_elements * 4;
+
+        // Read witness elements.
         loop {
+            // Check if we're done processing all elements.
+            if self.element_idx >= witness_elements {
+                return Ok(false);
+            }
+
             if bytes.is_empty() {
                 return Ok(true);
             }
 
-            if self.element_length_prefix_read {
-                let v = self.buffer.get_mut(self.idx).expect("we created this last call");
-                let copy_len = bytes.len().min(self.bytes_to_read);
+            // If we have some bytes to read, then reading element data.
+            // Else we are reading the element's length.
+            if let Some(bytes_to_read) = self.element_bytes_remaining {
+                let copy_len = bytes.len().min(bytes_to_read);
 
-                v.extend_from_slice(&bytes[..copy_len]);
-                *bytes = &bytes[copy_len..];
-                self.bytes_to_read -= copy_len;
+                // Ensure we have enough space.
+                let required_len = self.cursor + copy_len;
+                if required_len > self.content.len() {
+                    self.resize_if_needed(required_len);
+                }
 
-                if self.bytes_to_read == 0 {
-                    self.element_length_prefix_read = false;
-                    self.idx += 1;
-                    if self.idx == self.witness_elements {
-                        return Ok(false);
-                    }
+                self.content[self.cursor..self.cursor + copy_len]
+                    .copy_from_slice(&bytes[..copy_len]);
+                self.cursor += copy_len;
+                *bytes = &bytes[copy_len..];
+                let remaining = bytes_to_read - copy_len;
+
+                if remaining == 0 {
+                    // Element complete, move to next element.
+                    self.element_idx += 1;
+                    self.element_length_decoder = CompactSizeDecoder::new();
```

Comment:
I don't think we technically need this line since `mem::take` puts a default decoder in here already. But maybe this is easier to read like this.